### PR TITLE
eXtendable Output Functions as first-class citizen

### PR DESCRIPTION
**Pull request by reneme**
### Motivation

Most PQC schemes use XOFs as a building block in one or the other way. This is a proposal to establish `XOF` as a new public API base class. The only implementation currently provided is `SHAKE-128` and `SHAKE-256`.

### Pull Request Dependencies

* ~#3673~

### Find included

* New top-level module "xof" with `class XOF` as abstract base class
  * `start()` allows priming certain XOFs with a `salt` and `key` (otherwise its a NOOP) 
  * `update()` allows streamed input of an arbitrarily long message
  * `output()` allows streamed access to the XOF's infinitely long output stream
  * typical implementations won't allow `update()`s after the first call to `output()`
* `SHAKE_128_XOF` and `SHAKE_256_XOF` as sub-classes of `XOF`
* `cSHAKE_128_XOF` and `cSHAKE_256_XOF` as sub-classes of `XOF` (for internal use only)
* internal helper functions (roughly) as defined in [NIST SP.800-185](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf) Section 2.3
  (those functions can be reused to implement KMAC -- see #3570)

#### XOFs based on stream ciphers

Kyber and Dilithium build their "90s-crypto" variants on XOF constructions based on AES-256/CTR. We could look into providing support for such stream-cipher XOFs independent of the CRYSTALS algorithms. (See #3672)

#### Retrofitting PQC algorithms

That's future work and will be done in a follow-up. (See #3672)

### CODE REVIEW

**Comment by reneme** [comment]
Rebased onto @falko-strenzke's Keccak refactoring as a sanity checks. Findings are in #3673.

File: src/lib/base/buf_input.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+* internal state with an arbitrary number of streamlined update() calls.
+*/
+class BOTAN_PUBLIC_API(3, 2) Buffered_Input {
+   public:
+      /**
+      * Add new input to process.
+      * @param in the input to process as a byte array
+      * @param length of param in in bytes
+      */
+      void update(const uint8_t in[], size_t length) { add_data(in, length); }
```
Comment (randombit) [inline src/lib/base/buf_input.h:31]: `Buffered_Computation` is already deprecated, I don't think we should use it for the XOF

This also allows some modernization and minimization of the API. In fact we could literally go to just taking a `std::span<const uint8_t>` and be done with it.

File: src/lib/hash/shake/shake.h [hunk 2]
```diff
@@ -9,71 +10,70 @@
...
+}  // namespace detail
+
 /**
-* SHAKE-256
+* SHAKE-128 with a fixed output length
+*
+* For variable output lengths consider using the XOF interface
 */
-class SHAKE_256 final : public HashFunction {
+class SHAKE_128 final : public detail::SHAKE_Adapter<SHAKE_128_XOF, SHAKE_128> {
```
Comment (randombit) [inline src/lib/hash/shake/shake.h:64]: All of this seems completely unnecessary. Why not just hold a `std::unique_ptr<XOF>` of the appropriate type?

File: src/lib/stream/shake_cipher/shake_cipher.h [hunk 3]
```diff
@@ -1,77 +1,128 @@
...
-      void set_iv_bytes(const uint8_t iv[], size_t iv_len) final;
-
-   protected:
-      size_t m_shake_rate;
+      void cipher_bytes(const uint8_t in[], uint8_t out[], size_t length) final {
+         assert_key_material_set();
+
+         const auto block_size = m_keystream_buffer.size();
+
+         auto cipher_some = [&](size_t bytes) {
```
Comment (randombit) [inline src/lib/stream/shake_cipher/shake_cipher.h:65]: I don't see why this logic has to be in a header

File: src/lib/base/buf_input.h (hunk 1 above)
Comment (reneme) [reply src/lib/base/buf_input.h:31]: Mhm, fair enough. Albeit, I found it quite handy to have `update_le()`, for instance.

You're proposing to omit the common base interface altogether, right?

**Comment by reneme** [comment]
The whole Keccak-Refactoring is becoming a mess with several inter-dependent pull requests. Sorry for that. Lets put this one on-hold until #3673 and #3675 are integrated.

File: src/lib/stream/shake_cipher/shake_cipher.h (hunk 3 above)
Comment (reneme) [reply src/lib/stream/shake_cipher/shake_cipher.h]: Yeah, that was still sketchy. Addressed in #3675.

Now that `Keccak_Permutation::squeeze()` can handle multiple calls to pull the sponge output in arbitrarily-sized chunks, I decided not the base `SHAKE_Cipher` on the `SHAKE_XOF` but directly on `Keccak_Permutation`. There's no difference in level of abstraction anyway.

File: src/lib/hash/shake/shake.h (hunk 2 above)
Comment (reneme) [reply src/lib/hash/shake/shake.h]: The rationale is similar to [the one discussed back in May](https://github.com/randombit/botan/pull/3525#discussion_r1186276249). I think, we never really came to a conclusion whether this is just a code smell or might cause actual trouble.

Regardless: I decided not to touch the `HashFunction` implementation of SHAKE in this PR at all. With #3675 introducing `Keccak_Permutation::squeeze()`, there's no significant difference in level of abstraction between `Keccak_Permutation` and `SHAKE_XOF`. Hence, I think it's just fine to avoid the indirection via `SHAKE_XOF` here.

**Comment by reneme** [comment]
Re: XOFs based on stream ciphers

Maybe we should step back and rethink the XOF interface somewhat. Looking at [NIST SP.800-185](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf) they define things like KMAC-XOF that is KMAC with a variable output length. I.e. just like the AES/CTR-XOF used in CRYSTALS Kyber/Dilithium, it takes a key (and potentially other customization parameters).

I think it would be great to incorporate such XOFs off the bat, when introducing a new API.

**Comment by reneme** [comment]
To my mind an abstract XOF should have two phases:

1. Parameterization
2. Output

The "output phase" is simple and inherently generic. The user just calls `::output()` as often as they want to read the indefinite stream of XOF output bytes.

The "parameterization phase" is algorithm-specific, unfortunately. Here are some relevant examples:

| Algorithm   | Parameters                                |
| ----------- | ----------------------------------------- |
| SHAKE       | an arbitrary-length message `msg`         |
| cSHAKE      | arbitrary-lengths `N` and `S`, and `msg`  |
| AES/CTR-XOF | a short `iv` and `key`                    |
| KMAC-XOF    | arbitrary-length `key`, `msg` and `salt`  |

Reflecting this parameter space in a common `XOF` base-interface is possible but probably won't result in a convenient API. Except, if the only "common" feature of this interface is the `::output()` method which could be shared by all XOFs.

**Comment by reneme** [comment]
Given the above-described constraints, it seems reasonable to abandon the ubiquitous `::create("AlgoSpec")` factory methods for XOFs. Instead, the user could instantiate the concrete XOF subclass, parameterize it and then (maybe) use it via the abstract `XOF` base class that just provides the `::output()` functionality.

@randombit Does that sound like a reasonable abstraction to you?

**Comment by reneme** [comment]
Benchmarking the SHAKE-128/256 XOF showed comparable results for `update()` and `output()` and correspond to [the measured performance of SHA-3(256)/SHA-3(512)](https://github.com/randombit/botan/pull/3673#issuecomment-1691130234). No obvious bottlenecks.

|            | absorb()    | squeeze()   |
| ---------- | ----------- | ----------- |
| SHAKE-128  | 555MiB/sec  | 530MiB/sec  |
| SHAKE-256  | 448MiB/sec  | 432MiB/sec  |

**Comment by randombit** [comment]
>  Instead, the user could instantiate the concrete XOF subclass, parameterize it and then (maybe) use it via the abstract XOF base class that just provides the ::output() functionality.

I very much dislike this. A lot of effort was spent in 2->3 in removing all of the concrete subclasses from the public API; this minimizes API and ABI surface. This would also complicate the FFI interface.

I don't think CTR mode "XOF" is worth optimizing for. It is a weird corner case that should not be exposed to applications. And it seems Kyber is dropping "90s mode" entirely (#3543) so it is probably going to be removed soonish anyway.

Then we're left with 

* SHAKE: arbitrary length msg
* cSHAKE: arbitrary length msg. `N` is not for us - quoting SP 800-185 "N is a function-name bit string, used **by NIST** to define functions based on cSHAKE. When no function other than cSHAKE is desired, N is set to the empty string." and so can be ignored. And `S` is a domain separator and can be instantiated in the usual way eg `cSHAKE(256, my_domain_dep)`
* KMAC-XOF: arbitrary length msg, key, and salt.

```
class XOF {
  virtual void start(std::span<const uint8_t> salt = {}, std::span<uint8_t> key = {}) = 0;
  virtual bool valid_salt_length(size_t salt_len) = 0;
  // returns Key_Length_Specification(0) if key is not supported
  virtual Key_Length_Specification key_spec() const = 0;
  virtual void update(std::span<const uint8_t> input);
  //...
};
```

CTR, which doesn't support inputs, is handled by the funny case that `XOF::update` throws `Not_Implemented`.

**Comment by reneme** [comment]
@randombit Thanks for your opinion. Your points make sense and I agree in general. I'll be AFK until the beginning of September and will come back to this after.

**Comment by randombit** [comment]
Needs a rebase now that the SHA-3 refactor is merged; I'll review after.

**Comment by reneme** [comment]
> Needs a rebase now that the SHA-3 refactor is merged; I'll review after.

Rebased. I hope, I get time today to work on the extended API [discussed here](https://github.com/randombit/botan/pull/3671#issuecomment-1691595134).

File: src/tests/data/xof/shake.vec [hunk 4]
```diff
@@ -0,0 +1,539 @@
+# These vectors are selected from the NIST CAVS file for SHAKE
+#  Generated on Thu Jan 28 14:46:49 2016
+
+[SHAKE-128]
```
Comment (reneme) [inline src/tests/data/xof/shake.vec:4]: Note to self: the Keccak permutation provides a `bmi2` implementation. Probably, we want to insert

```
#test cpuid bmi2
```

in that test vector file to exercise the alternative provider.

File: src/lib/base/buf_input.h (hunk 1 above)
Comment (randombit) [reply src/lib/base/buf_input.h:31]: Exactly so, yes

File: src/lib/xof/shake_xof/shake_xof.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+SHAKE_XOF::SHAKE_XOF(size_t capacity) : m_keccak(capacity, 0xF, 4), m_output_generated(false) {
+   BOTAN_ASSERT_NOMSG(m_keccak.byte_rate() >= 72 && m_keccak.byte_rate() <= 168);
+}
+
+void SHAKE_XOF::clear() {
+   m_keccak.clear();
+   m_output_generated = false;
+}
+
+void SHAKE_XOF::add_data(const uint8_t key[], size_t length) {
```
Comment (randombit) [inline src/lib/xof/shake_xof/shake_xof.cpp:25]: Calling this param `key` seems odd since there is no requirement that XOF input be secret, and is inconsistent with the other declarations which call this `input` which seems a better name to capture the non-specific nature of the bytes here.

Comment (reneme) [reply src/lib/xof/shake_xof/shake_xof.cpp:25]: For sure. Not sure where this came from. Perhaps a copy & paste quirk. Good catch.

**Comment by reneme** [comment]
For tomorrow:

* [ ] Revert the `Buffered_Input` refactoring (see https://github.com/randombit/botan/pull/3671#discussion_r1297073573)
* [ ] Rebase #3672 and try the extended API (`XOF::start()`)
* [ ] Look into cSHAKE

File: src/lib/xof/xof.h [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+      void output(std::span<uint8_t> output) { generate_bytes(output); }
+
+      /**
+       * @return the next single output byte
+       */
+      uint8_t output_next_byte() {
+         uint8_t out;
+         generate_bytes({&out, 1});
+         return out;
+      }
```
Comment (reneme) [inline src/lib/xof/xof.h:165]: `XOF` does not inherit from a generic `Buffered_Input` class anymore. Instead it directly defines `update(std::span<>)` and no additional convenience methods. For consistency, we should probably remove the convenience methods of `output()` as well, no?

My two cents: I did enjoy having some convenience overloads (particularly taking `std::string_view` and Big/Little-Endian integers). Re-implementing them here as well as later in `HashFunction` and `MessageAuthenticationCode` feels like violating the DRY principle, though. Hence, I wouldn't do that.

Nevertheless, I'd suggest keeping two flavors of `output()`:

* `void output(std::span<>)`
   to allow the user to handle memory allocation
* `T output(size_t bytes)` with `T` being a user-defined container that defaults to `secure_vector<>`
   to let the user conveniently pull data wrapped in their favorite vector-ish.

**Review by randombit — COMMENTED** [review]
First pass review

File: src/lib/xof/shake_xof/shake_xof.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+void SHAKE_XOF::start(std::span<const uint8_t> salt, std::span<const uint8_t> key) {
+   BOTAN_UNUSED(key);
+   BOTAN_STATE_CHECK(!m_output_generated);
+
+   if(!m_function_name.empty() || !salt.empty()) {
+      // We deliberately do not check the salt length here. There's no technical
+      // reason to reject a salt buffer passed in by an application that exceeds
+      // the (arbitrarily set) valid_salt_length() restriction.
+      keccak_absorb_padded_strings_encoding(*this, block_size(), m_function_name, salt);
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/xof/shake_xof/shake_xof.cpp:64]: Yikes, indeed! The condition above was previously meant to exclude non-cSHAKE instances of this base class. But I must have broken it in some restructuring. Anyway, that's certainly not optimal.

I'll look into giving the cSHAKE variants their own base class and leave the salt logic out of the vanilla SHAKE implementation.

File: src/lib/xof/shake_xof/shake_xof.cpp [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+cSHAKE_128_XOF::cSHAKE_128_XOF(std::vector<uint8_t> function_name) :
+      SHAKE_XOF(256, std::move(function_name), 0b00, 2) {}
+
+cSHAKE_128_XOF::cSHAKE_128_XOF(std::span<const uint8_t> function_name) :
+      cSHAKE_128_XOF(std::vector<uint8_t>{function_name.begin(), function_name.end()}) {}
+
+cSHAKE_128_XOF::cSHAKE_128_XOF(std::string_view function_name) :
+      cSHAKE_128_XOF(std::vector<uint8_t>{cast_char_ptr_to_uint8(function_name.data()),
+                                          cast_char_ptr_to_uint8(function_name.data()) + function_name.size()}) {}
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/xof/shake_xof/shake_xof.cpp:106]: I'd love that. Unfortunately, the base-class constructor additionally takes the `capacity`. This value is hard-coded in the concrete overloads in cSHAKE-128/256 classes. Therefore we can't just conveniently use the delegates in the derived classes. 😞

[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/xof/shake_xof/shake_xof.cpp:106]: Nevertheless, moving the data type conversions into the base class helps a little bit. Then the derived class constructors simply delegate to their base-class counter part.

File: src/lib/xof/xof.h [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+       */
+      virtual bool valid_salt_length(size_t salt_len) const = 0;
+
+      /**
+       * @returns an object describing limits on the key size
+       */
+      virtual Key_Length_Specification key_spec() const = 0;
+
+      /**
+       * @return the intrinsic processing block size of this XOF
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/xof/xof.h:101]: I changed this to be a pure-virtual, forcing downstream implementations to make up their minds.

File: src/tests/test_xof.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+            const std::string provider(xof->provider());
+            result.test_is_nonempty("provider", provider);
+            result.test_eq(provider, xof->name(), algo);
+
+            // Some XOFs don't accept input at all. We assume that this stays the same
+            // after calling `XOF::clear()`.
+            const auto new_accepts_input = xof->accepts_input();
+
+            result.confirm("advertised block size is > 0", xof->block_size() > 0);
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/tests/test_xof.cpp:55]: [Yes](https://github.com/randombit/botan/pull/3671#discussion_r1321161215).

File: src/lib/xof/xof.h [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+      /**
+       * @return the hash function name
+       */
+      virtual std::string name() const = 0;
+
+      /**
+       * Some XOFs can be parameterized with a @p salt and/or @p key. If required,
+       * this must be called before calling XOF::update(). XOFs that don't support
+       * either of these parameters will simply ignore it silently.
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/xof/xof.h:74]: Changed as described [here](https://github.com/randombit/botan/pull/3671#discussion_r1321161215).

File: src/lib/permutations/keccak_perm/keccak_helpers.cpp [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <botan/internal/bit_ops.h>
+#include <botan/internal/loadstor.h>
+
+namespace Botan {
+
+namespace {
+
+size_t int_encoding_size(uint64_t x) {
+   return ceil_tobytes(std::max(uint8_t(1), ceil_log2(x + 1)));
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/permutations/keccak_perm/keccak_helpers.cpp:22]: True, that'll result in a wrong result. The int-encodings are mostly used for encoding bitstring-lengths (for salts, function names and friends). I.e. in practice that shouldn't ever happen. I added an assert to be on the safe side, though. Good catch!

**Comment by reneme** [comment]
Thanks for the review! I rebased to latest master (after merging the `std::span` refactorings) and addressed your review comments.

Most notably, I split the `SHAKE_XOF` and `cSHAKE_XOF` into two separate base classes (and modules). That results in a handful duplicated lines (`::add_data()`, `::generate_bytes()`) but splits off the salt handling in `::start()` much more logically.

**Review by randombit — COMMENTED** [review]
Looks good now. Should we have some tests for cSHAKE as well?

**Comment by falko-strenzke** [comment]
There are four [test vectors from NIST](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/cSHAKE_samples.pdf), all of which have an empty value for N.

I just implemented cSHAKE for another libgcrpyt and created two test vectors with non empty N using  https://asecuritysite.com/golang/cs. I checked their results for non-empty N and that is fine.

```
  { /* Created with https://asecuritysite.com/golang/cs */
    GCRY_MD_CSHAKE128,
    "00010203",
    "ABC",
    "Email Signature",
    32,
    "5CF74DC523ADC0B97EC3614E703835277E9F818879AA1EAE5B2B4E4472EB6A68" },
  { /* Created with https://asecuritysite.com/golang/cs */
   CSHAKE256,
    "00010203", // data hex
    "ABC", // N as C string
    "Email Signature", // S as C string
    32, // output size in bytes
    "0C34C14C4A56E5FC01BE8C04C759DA61437E86B88DF3E21A934436D427A85E9D"  // expected output hex
}, 
```

**Comment by randombit** [comment]
Thanks for checking this @falko-strenzke. @reneme can we get these included here so we have some baseline check on cSHAKE? I guess it is a little complicated by cSHAKE not being exposed to applications so you can't immediately use the normal XOF test.

**Comment by reneme** [comment]
@falko-strenzke Nice! Thanks for the test vectors. I did look for something "official" for cSHAKE but wasn't successful.

@randombit I hacked the XOF allocation in `test_xof.cpp` so that it explicitly deals with the internal cSHAKE XOFs when needed. Similarly, I also added vectors for the AES/CTR-XOF in #3672 a while ago.

**Comment by randombit** [comment]
@reneme looks good thanks. I couldn't find the tests themselves though - are you missing a `git add` of `cshake.vec`?

**Comment by reneme** [comment]
I was fighting some linker visibility issues. Had to move the `cSHAKE_XOF::provider()` and `::block_size()` implementations into the *.cpp file to avoid exposing the `Keccak_Permutation` in the shared library.

Now everything should turn green. ☺️

**Review by randombit — APPROVED** [review]

**Comment by reneme** [comment]
# 🥳
